src/Tokenomia/Vesting/PrivateSale.hs

```diff
@@ -0,0 +1,223 @@
+    , _psStart :: POSIXTime
+    , _psTranches :: [Tranche]
+    , _psAssetClass :: Value.AssetClass -- Vesting token
+    , _psInvestors :: [PrivateInvestor]
+    }
+    deriving stock (Generic, Show)
+    deriving anyclass (FromJSON)
+
+data Tranche = Tranche
+    { _tranchePercentage :: Integer
+    , _trancheDuration :: Integer
+    }
+    deriving stock (Generic, Show)
+    deriving anyclass (FromJSON)
+
+data PrivateInvestor = PrivateInvestor
+    { _piAddress :: B.Address
+    , _piAllocation :: Integer -- Amount of vesting tokens to lock (note, this is likely NOT equal to the total amounts of investments, as different tokens)
+    , _piInvestments :: [Investment]
+    }
+    deriving stock (Generic, Show)
+    deriving anyclass (FromJSON)
+
+data Investment = Investment
+    { _invTx :: B.TxHash -- Tx that sends below asset class to payment address
+    , _invAssetClass :: Value.AssetClass
+    , _invAmount :: Integer -- Amount of above asset class expected to be sent to payment address
+    }
+    deriving stock (Generic, Show)
+    deriving anyclass (FromJSON)
+
+makeLenses ''PrivateSale
+makeLenses ''Tranche
+makeLenses ''PrivateInvestor
+makeLenses ''Investment
+
+verifyPrivateSale ::
+    ( MonadIO m
+    , MonadError TokenomiaError m
+    , MonadReader Environment m
+    ) =>
+    m ()
+verifyPrivateSale = do
+    liftIO . putStrLn $ "Please enter a filepath with JSON data"
+    jsonFilePath <- liftIO getLine
+    verifyPrivateSale' jsonFilePath
+
+verifyPrivateSale' ::
+    ( MonadIO m
+    , MonadError TokenomiaError m
+    , MonadReader Environment m
+    ) =>
+    FilePath ->
+    m ()
+verifyPrivateSale' jsonFilePath = do
+    ps <- jsonToPrivateSale jsonFilePath
+    treasAddrTxs <- getTreasAddrTxs ps -- :: [B.AddressTransaction] <- m [B.AddressTransaction]
+    let invTxhs = getTxhsByPrivateSale ps
+        txhs = verifyTxHashList invTxhs treasAddrTxs
+    if length txhs == length invTxhs
+        then do
+            verifyTxs ps txhs
+        else throwError . BlockFrostError . B.BlockfrostError $ "Missing Transactions"
+
+jsonToPrivateSale ::
+    ( MonadIO m
+    , MonadError TokenomiaError m
+    , MonadReader Environment m
+    ) =>
+    FilePath ->
+    m PrivateSale
+jsonToPrivateSale jsonFilePath = do
+    eitherErrPs <- liftIO . eitherDecodeFileStrict $ jsonFilePath
+    liftEither (first (BlockFrostError . B.BlockfrostError . pack) eitherErrPs)
+
+getTreasAddrTxs ::
+    ( MonadIO m
+    , MonadError TokenomiaError m
+    , MonadReader Environment m
+    ) =>
+    PrivateSale ->
+    m [B.AddressTransaction]
+getTreasAddrTxs ps = do
+    prj <- projectFromEnv''
+    eitherErrAddrTxs <- liftIO $ B.runBlockfrost prj (B.getAddressTransactions (ps ^. psAddress))
+    liftEither $ first BlockFrostError eitherErrAddrTxs
+
+getTxhsByPrivateSale :: PrivateSale -> [B.TxHash]
+getTxhsByPrivateSale ps = invTxhs
+  where
+    invTxhs :: [B.TxHash]
+    invTxhs = (^. invTx) <$> invs
```

Comment:
No reason to give this the name if its the entire return. To be honest, I dont think we need a `where` on this at all
```hs
(^. invTx) <$> getPsInvestments ps
```